src/Ocelot/DownstreamRouteFinder/Finder/DownstreamRouteFinder.cs

```diff
@@ -47,10 +53,33 @@ public Response<DownstreamRouteHolder> Get(string upstreamUrlPath, string upstre
             return new ErrorResponse<DownstreamRouteHolder>(new UnableToFindDownstreamRouteError(upstreamUrlPath, httpMethod));
         }
 
-        private static bool RouteIsApplicableToThisRequest(Route route, string httpMethod, string upstreamHost)
+        private static bool RouteIsApplicableToThisRequest(Route route, string httpMethod, string upstreamHost, IHeaderDictionary requestHeaders)
         {
-            return (route.UpstreamHttpMethod.Count == 0 || route.UpstreamHttpMethod.Select(x => x.Method.ToLower()).Contains(httpMethod.ToLower())) &&
-                   (string.IsNullOrEmpty(route.UpstreamHost) || route.UpstreamHost == upstreamHost);
+            return (route.UpstreamHttpMethod.Count == 0 || RouteHasHttpMethod(route, httpMethod)) &&
+                   (string.IsNullOrEmpty(route.UpstreamHost) || route.UpstreamHost == upstreamHost) &&
+                   (route.UpstreamHeaderRoutingOptions == null || !route.UpstreamHeaderRoutingOptions.Enabled() || RouteHasRequiredUpstreamHeaders(route, requestHeaders));
+        }
+
+        private static bool RouteHasHttpMethod(Route route, string httpMethod)
+        {
+            httpMethod = httpMethod.ToLower();
+            return route.UpstreamHttpMethod.Select(x => x.Method.ToLower()).Contains(httpMethod);

```

Comment:
```suggestion
            var method = new HttpMethod(httpMethod);
            return route.UpstreamHttpMethod.Contains(method);
```

Reply:
From [line 65](commits/6d29040db6570a01847ef596a2cdc583e146ad6c#diff-9c51794162197f4d2f9ad8f40e3155d9a4228b32e09b8cc76d5e10f77e6682cbR65):
`route.UpstreamHttpMethod.Select(x => x.Method).Contains(httpMethod, StringComparer.OrdinalIgnoreCase)`
`route.UpstreamHttpMethod.Contains(new HttpMethod(httpMethod))`
Which one is shorter?

Another issue is transforming collection of `HttpMethod` to a collection of `String`, and checks method string finally.
Doesn't it strange? What is performance, memory, etc.?

Reply:
**Fixed** in https://github.com/ThreeMammals/Ocelot/pull/1684/commits/f372ea9783fd1d2fe2496e41d58b16f018aedded